rust/experimental/server/src/store/memory.rs

```diff
@@ -413,6 +413,18 @@ impl Store for MemoryStore {
                                 last_block_id = -1;
                             }
                         }
+
+                        // get block_ids filter
+                        // In AQE, after executing the sub-QueryStages, collect the shuffle data size
+                        // So if we can filter block, it will improve the performance of AQE.
+                        candidate_blocks = candidate_blocks
```

Comment:
This part looks not correct. I hope this could be bound to the `read_partial_data_with_max_size_limit` .

Without the bitmap filter, we will read it the (last_block_id, end] without the limited size. But for now, I think it will read from (last_block_id, end] with filter and then in a limited size. Right?

So it's better to refactor the `read_partial_data_with_max_size_limit`, maybe the name of `read_partial_data_with_max_size_limit_and_filter` will be better. Like that

```
pub(crate) fn read_partial_data_with_max_size_limit_and_task_filter<'a>(
        &'a self,
        blocks: Vec<&'a PartitionedDataBlock>,
        fetched_size_limit: i64,
        serialized_expected_task_ids_bitmap: Option<Treemap>,
    ) -> (Vec<&PartitionedDataBlock>, i64) {
        let mut fetched = vec![];
        let mut fetched_size = 0;

        for block in blocks {
            if &serialized_expected_task_ids_bitmap.is_some() {
                if !&serialized_expected_task_ids_bitmap.unwrap().contains(block.task_attempt_id as u64) {
                    continue;
                }
            }
            if fetched_size >= fetched_size_limit {
                break;
            }
            fetched_size += block.length as i64;
            fetched.push(block);
        }

        (fetched, fetched_size)
    }
```

Reply:
Your consideration is correct. 
Integrating filter logic into read_partial_data_with_max_size_limit_and_task_filter functions is indeed a better design choice, so that the filter can be applied directly when the data is read, rather than filtering and then reading. 
The advantage of this is that data can be processed more efficiently, especially when the amount of data is large or the filtering conditions are complex.
Thank you for your meticulous guidance and patience.